pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -148,15 +148,36 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
 			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
 			// Return and don't requeue
 			reqLogger.Info(fmt.Sprintf("NodeMaintenance Object: %s Deleted ", request.NamespacedName))
-			maintanenceMode = false
 		} else {
 			// Error reading the object - requeue the request.
 			reqLogger.Info("Error reading the request object, requeuing.")
 			return reconcile.Result{}, err
 		}
 	}
 
-	nodeName := request.Name
+	//Add finalizer when object is created
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !containsString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer) {
+			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer)
+			if err := r.client.Update(context.TODO(), instance); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if containsString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer) {
+			// remove our finalizer from the list and update it.
+			instance.ObjectMeta.Finalizers = removeString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer)
+			if err := r.client.Update(context.Background(), instance); err != nil {
```

Comment:
You will have to update the operator status after you uncordoned the node. Otherwise you would not retry to uncordon the node if you have an error when calling the api after this call. Normally it is like this:

1. check if I have something to do
2. do it
3. update the status of the main CR

This way you will retry in case you have any kind of errors until you reached your target. If you do it this way:

1. check if I have something to do
2. update the status of the main CR
3. do it

you would never retry 3. if it fails the first time because you already updated the status.